compose/legacy.py

```diff
@@ -72,26 +72,31 @@ def add_labels(project, container, name):
 def migrate_project_to_labels(project):
     log.info("Running migration to labels for project %s", project.name)
 
-    client = project.client
-    for container in client.containers(all=True):
-        name = get_container_name(container)
-        if not is_valid_name(name):
-            continue
-        add_labels(project, Container.from_ps(client, container), name)
+    containers = get_legacy_containers(
+        project.client,
+        project.name,
+        project.service_names,
+        stopped=True,
+        one_off=False)
 
+    for container in containers:
+        add_labels(project, container)
 
-def get_legacy_container_names(
+
+def get_legacy_containers(
         client,
         project,
         services,
         stopped=False,
         one_off=False):
 
-    for container in client.containers(all=stopped):
-        name = get_container_name(container)
-        for service in services:
+    containers = client.containers(all=stopped)
+
+    for service in services:
+        for container in containers:
```

Comment:
Yes, sorry, the actual *fix* is a bit subtle - it relies on the fact that `Project` keeps its services in order, and so `project.service_names` returns them in order. That's a widely relied-upon constraint, though, so I don't think this is brittle.